youtube_dl/extractor/adn.py

```diff
@@ -119,59 +124,77 @@ def _get_subtitles(self, sub_path, video_id):
+            config_url, video_id,
+            'Downloading player config JSON metadata')['player']['options']
+
+        user = player_config['user']
+        if not user.get('hasAccess'):
+            raise ExtractorError('This video is only available for paying users')
+            # self.raise_login_required() # FIXME: Login is not implemented
+
+        token = self._download_json(
+            user.get('refreshTokenUrl') or urljoin(self._API_BASE_URL, '/player/refresh/token'),
+            video_id, 'Downloading access token', headers={'x-player-refresh-token': user['refreshToken']},
+            data=b'')['token']
+
+        links_url = player_config.get('videoUrl') or urljoin(self._API_BASE_URL, '/player/video/%s/link' % video_id)
+        self._K = ''.join([random.choice('0123456789abcdef') for _ in range(16)])
+        message = bytes_to_intlist(json.dumps({
+            'k': self._K,
+            't': token,
+        }))
+
+        # Sometimes authentication fails for no good reason, retry with
+        # a different random padding
+        links_data = None
+        for _ in range(3):
             padded_message = intlist_to_bytes(pkcs1pad(message, 128))
             n, e = self._RSA_KEY
             encrypted_message = long_to_bytes(pow(bytes_to_long(padded_message), e, n))
             authorization = base64.b64encode(encrypted_message).decode()
-            links_data = self._download_json(
-                urljoin(self._BASE_URL, links_url), video_id,
-                'Downloading links JSON metadata', headers={
-                    'Authorization': 'Bearer ' + authorization,
-                })
-            links = links_data.get('links') or {}
-            metas = metas or links_data.get('meta') or {}
-            sub_path = sub_path or links_data.get('subtitles') or \
-                'index.php?option=com_vodapi&task=subtitles.getJSON&format=json&id=' + video_id
-            sub_path += '&token=' + token
-            error = links_data.get('error')
-        title = metas.get('title') or video_info['title']
+
+            try:
+                links_data = self._download_json(
+                    urljoin(self._BASE_URL, links_url), video_id,
+                    'Downloading links JSON metadata', headers={
+                        'X-Player-Token': authorization
+                    },
+                    query={
+                        'freeWithAds': 'true',
+                        'adaptive': 'false',
+                        'withMetadata': 'true',
+                        'source': 'Web'
+                    }
+                )
+                break
+            except ExtractorError as e:
+                if not isinstance(e.cause, compat_HTTPError):
+                    raise e
+
+                if e.cause.code == 401:
+                    # This usually goes away with a different random pkcs1pad, so retry
+                    continue
+
+                error = self._parse_json(e.cause.fp.read(), video_id)
```

Comment:
- you can just do `e.cause.read()` instead of `e.cause.fp.read()`.
- extract `message` into a variable as it's going to be used either way.